core/payments/ondemand/on_demand_vault_monitor.go

```diff
@@ -0,0 +1,114 @@
+package ondemand
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/big"
+	"time"
+
+	"github.com/Layr-Labs/eigenda/core/payments"
+	"github.com/Layr-Labs/eigensdk-go/logging"
+	gethcommon "github.com/ethereum/go-ethereum/common"
+)
+
+// Checks for updates to the PaymentVault contract, and updates ledgers with the new state
+type OnDemandVaultMonitor struct {
+	logger logging.Logger
+	// fetches data from the PaymentVault
+	paymentVault payments.PaymentVault
+	// how frequently to fetch state from the PaymentVault to check for updates
+	updateInterval time.Duration
+	// function to get accounts that need to be updated
+	getAccountsToUpdate func() []gethcommon.Address
+	// function to update the total deposit for an account
+	updateTotalDeposit func(accountID gethcommon.Address, newTotalDeposit *big.Int) error
+	// cancels the periodic update routine
+	cancelFunc context.CancelFunc
+}
+
+// Creates a new OnDemandVaultMonitor and starts a routine to periodically check for updates
+func NewOnDemandVaultMonitor(
+	ctx context.Context,
+	logger logging.Logger,
+	paymentVault payments.PaymentVault,
+	updateInterval time.Duration,
+	getAccountsToUpdate func() []gethcommon.Address,
+	updateTotalDeposit func(accountID gethcommon.Address, newTotalDeposit *big.Int) error,
+) (*OnDemandVaultMonitor, error) {
+	if updateInterval <= 0 {
+		return nil, errors.New("updateInterval must be > 0")
+	}
+
+	ctxWithCancel, cancel := context.WithCancel(ctx)
+
+	monitor := &OnDemandVaultMonitor{
+		logger:              logger,
+		paymentVault:        paymentVault,
+		updateInterval:      updateInterval,
+		getAccountsToUpdate: getAccountsToUpdate,
+		updateTotalDeposit:  updateTotalDeposit,
+		cancelFunc:          cancel,
+	}
+
+	go monitor.runUpdateLoop(ctxWithCancel)
+	return monitor, nil
+}
+
+func (vm *OnDemandVaultMonitor) Stop() {
+	vm.cancelFunc()
+}
```

Comment:
unimportant nit: isn't it simpler to just pass in a cancellable context to the constructor directly?